File: KafkaManagerUI_SpringBoot/src/main/java/com/kafkaui/entity/TopicRequest.java
package com.kafkaui.entity;

import jakarta.persistence.*;
import lombok.Data;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;

@Entity
@Table(name = "topic_requests")
@Data
public class TopicRequest {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(nullable = false)
    private String username;

    @Column(nullable = false)
    private String operation;

    @Column(name = "topic_name", nullable = false)
    private String topicName;

    @Column(name = "num_partitions")
    private Integer numPartitions;

    @Column(name = "replication_factor")
    private Integer replicationFactor;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private String config;

    @Column(nullable = false)
    private String cluster = "default";

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "status", nullable = false)
    private RequestStatus status = RequestStatus.PENDING;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "approved_by")
    private String approvedBy;

    @Column(name = "admin_comment")
    private String adminComment;
}

File: backend/app/models/topic_request.py
from sqlalchemy import Column, Integer, String, DateTime, Enum as SQLEnum, JSON
from sqlalchemy.sql import func
from app.db.base import Base
from app.models.acl_request import RequestStatus  # Reuse the same enum

class TopicRequest(Base):
    __tablename__ = "topic_requests"

    id = Column(Integer, primary_key=True, index=True)
    username = Column(String, index=True, nullable=False)
    operation = Column(String, nullable=False)  # CREATE, ALTER, DELETE
    topic_name = Column(String, nullable=False, index=True)
    num_partitions = Column(Integer, nullable=True)  # For CREATE/ALTER
    replication_factor = Column(Integer, nullable=True)  # For CREATE only
    config = Column(JSON, nullable=True)  # Topic configuration as JSON
    cluster = Column(String, nullable=False, default="default")
    status = Column(SQLEnum(RequestStatus), default=RequestStatus.PENDING, nullable=False)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    approved_by = Column(String, nullable=True)  # Admin username
    admin_comment = Column(String, nullable=True)  # Admin comment on approve/reject
